### PR TITLE
Complete Flutter Testing Guide

### LINKED ISSUE #597: Adding Flutter Testing Guide
**Describe the solution you'd like**

Adding a Complete Flutter Testing Guide. 
I would like to write some sort of documentation on testing with Flutter so that Beginners will get a decent resource to start with.
Newcomers will get to know about the existing mocks, stubs that we already have and can use them.

**Approach to be followed**

1- What are tests and why do we need them ?
2- How to cover all the corner cases ?
3- How to start with Flutter testing?
4- What are mocks and to generate them?
5- How to stub methods and classes?
6- What to do if you find a bug while writing test for that file?

### PR DESCRIPTION
**Pull request by ALIRAZA17**
**What kind of change does this PR introduce?**

I have written a complete Flutter Testing Guide Document which explains everything from adding required packages to creating a test file and then running the test cases.

**Issue Number:**

Fixes #597 

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-docs/blob/develop/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by ALIRAZA17** [comment]
@palisadoes @literalEval @noman2002 I have added the Flutter Testing Guide. Can you please review it and if there's anything missing I can add?

**Comment by literalEval** [comment]
Awesome work @ALIRAZA17 :) 

I have two suggestions.

1. Add a better image.
2. Briefly elaborate on important parts of the testing system that Talawa has. The idea is that there are already a lot of tutorial on flutter testing. If we are writing one for ourselves, it should be customized to our needs. You can add things like Mocks for such and such classes already exist but all of the methods/attributes are not mocked for now etc.

**Comment by literalEval** [comment]
Please note that you should only follow these suggestions if @palisadoes agrees with them. Otherwise you are good to go.

**Comment by palisadoes** [comment]
@ALIRAZA17 I have done two Flutter courses and never written an app outside of it. Take the advice of @literalEval for now and let me know when you are done.

**Comment by ALIRAZA17** [comment]
> @ALIRAZA17 I have done two Flutter courses and never written an app outside of it. Take the advice of @literalEval for now and let me know when you are done.

ok ok Im on it. I'll let you know!!

**Comment by ALIRAZA17** [comment]
> Awesome work @ALIRAZA17 :)
> 
> I have two suggestions.
> 
> 1. Add a better image.
> 2. Briefly elaborate on important parts of the testing system that Talawa has. The idea is that there are already a lot of tutorial on flutter testing. If we are writing one for ourselves, it should be customized to our needs. You can add things like Mocks for such and such classes already exist but all of the methods/attributes are not mocked for now etc.

on it!

**Comment by ALIRAZA17** [comment]
@palisadoes @literalEval Can you please review it again I have updated the image and also added an example of test from talawa app. Is there anything I should remove or add in this doc?

**Comment by literalEval** [comment]
@ALIRAZA17 please remove the earlier image which you are not using now.

**Comment by ALIRAZA17** [comment]
> @ALIRAZA17 please remove the earlier image which you are not using now.

@literalEval done

**Comment by literalEval** [comment]
@palisadoes looks good to me. Just add the file in `sidebar.js`

**Review by Cioppolo14 — CHANGES_REQUESTED** [review]
Please edit sidebars.js to include this new page.

**Comment by ALIRAZA17** [comment]
> Please edit sidebars.js to include this new page. 

@Cioppolo14 is there something I have to edit?

**Comment by literalEval** [comment]
@ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

**Comment by ALIRAZA17** [comment]
> @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

Ok ok!!

**Comment by ALIRAZA17** [comment]
> @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.

@literalEval Added it can you check it if its ok?

**Comment by ALIRAZA17** [comment]
> > @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.
> 
> @literalEval @Cioppolo14 Added it can you check it if its ok?

**Comment by ALIRAZA17** [comment]
> > > @ALIRAZA17 open `sidebar.js` and add your filename in there. Open `sidebar.js` and you'll understand what to do.
> > 
> > 
> > @literalEval @Cioppolo14 @palisadoes Added it can you check it if its ok?

**Comment by literalEval** [comment]
Did it show up properly in the sidebar of your local build ?

**Comment by ALIRAZA17** [comment]
> Did it show up properly in the sidebar of your local build ?

@literalEval Now it is running There was a mistake from my side. Its good to go now

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please update the PR. 

1. The flutter testing guide is not visible in the sidebar. You need to correctly update sidebar.js to add the entry
2. The markdown file you added does not have the correct entries at the top like the other markdown files in the directory tree to be recognized by sidebar.js
3. Remove the image and references to it. It is not necessary.

**Comment by ALIRAZA17** [comment]
> Please update the PR.
> 
> 1. The flutter testing guide is not visible in the sidebar. You need to correctly update sidebar.js to add the entry
> 2. The markdown file you added does not have the correct entries at the top like the other markdown files in the directory tree to be recognized by sidebar.js
> 3. Remove the image and references to it. It is not necessary.

@palisadoes Im facing a problem of document id because my document id doesnt exist. Where to add that id of my document to configure it??

**Comment by palisadoes** [comment]
Look at the other files in the directory. You will see their IDs at the top

**Comment by ALIRAZA17** [comment]
image
When I add this in my document I face this error
image

**Comment by ALIRAZA17** [comment]
> image 

When I add this in my document I face this error 

image

**Comment by ALIRAZA17** [comment]
> Look at the other files in the directory. You will see their IDs at the top

Ive added it but it gave me that error I just showed you

**Comment by palisadoes** [comment]
Side bar does not check the file name
It checked the IDs of the files in the directory

**Comment by ALIRAZA17** [comment]
> Side bar does not check the file name It checked the IDs of the files in the directory

Ive updated the file and added the id but it gives me an error of invalid document id. If I remove that portion it works fine on my local machine and my document appears on the web page

**Comment by palisadoes** [comment]
Please follow the convention we have already. All files have IDs and the ID is referenced in the sidebar.

The first part of the entry in the sidebar is the directory tree and the last item of the entry is the Id.

Look at the other files in the other directories as a guide.

We need to be consistent or else the risk of breaking out documentation gets too high

**Comment by ALIRAZA17** [comment]
> Please follow the convention we have already. All files have IDs and the ID is referenced in the sidebar.
> 
> The first part of the entry in the sidebar is the directory tree and the last item of the entry is the Id.
> 
> Look at the other files in the other directories as a guide.
> 
> We need to be consistent or else the risk of breaking out documentation gets too high

@palisadoes Can you check it now the error is resolved

**Comment by palisadoes** [comment]
Fix this:

image

File: docs/developers/talawa/flutter_testing_guide.md [hunk 1]
```diff
@@ -0,0 +1,271 @@
...
+
+2- The `ChatService` class is instantiated, and the `sendMessageToDirectChat` method is called with the correct `chatId` and `messageContent` parameters.
+
+3- Finally, the test verifies that the `_chatMessageController` object has been updated with the correct `ChatMessage` object that was received from the mocked `GraphQL mutation` result.
+
+4- The `when` function is used to set up a mock behavior for the `gqlAuthMutation` method of the `_dbFunctions` object. The mocked behavior returns a `QueryResult` object that simulates the result of a GraphQL mutation. The `QueryResult` object contains a map with a key of 'sendMessageToDirectChat', which contains a value that represents the returned ChatMessage object from the mutation.
+
+Overall, this test verifies that the sendMessageToDirectChat method correctly triggers a GraphQL mutation and correctly handles the returned data by updating the _chatMessageController object with the expected ChatMessage object.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: _chatMessageController should be code

File: docs/developers/talawa/flutter_testing_guide.md [hunk 2]
```diff
@@ -0,0 +1,271 @@
...
+
+Here is a breakdown of what this test does
+
+1- The test starts by defining a mock object for the `_dbFunctions` class using the when function from the `Mockito` package. The mock object is set up to return a `QueryResult` object that simulates the result of a GraphQL mutation when the `gqlAuthMutation` method is called with the correct query and variables.
+
+2- The `ChatService` class is instantiated, and the `sendMessageToDirectChat` method is called with the correct `chatId` and `messageContent` parameters.
+
+3- Finally, the test verifies that the `_chatMessageController` object has been updated with the correct `ChatMessage` object that was received from the mocked `GraphQL mutation` result.
+
+4- The `when` function is used to set up a mock behavior for the `gqlAuthMutation` method of the `_dbFunctions` object. The mocked behavior returns a `QueryResult` object that simulates the result of a GraphQL mutation. The `QueryResult` object contains a map with a key of 'sendMessageToDirectChat', which contains a value that represents the returned ChatMessage object from the mutation.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: `sendMessageToDirectChat`

File: docs/developers/talawa/flutter_testing_guide.md [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+
+In Flutter, you can use stubs to replace real objects with mock objects or other simplified versions.
+
+To stub a method or class in Flutter, you can use the when() function provided by the `mockito` package. For example, if you have a method called myMethod() that you want to stub, you can do the following:
+
+```dart
+var myMock = MockMyObject();
+when(myMock.myMethod()).thenReturn('my result');
+```
+This will replace the `myMethod()` method on the `myMock` object with a stub that always returns the string 'my result'.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: `my result` should be code.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: `my result` should be code

File: docs/developers/talawa/flutter_testing_guide.md [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+ 4. Use the mock object in your test cases:
+
+    ```dart
+    test('my test case', () {
+    var myMock = MockMyObject();
+    // ... use the mock object ...
+    });
+    ```
+    
+   ### Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: This needs to be against the margin

File: docs/developers/talawa/flutter_testing_guide.md [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+  mock
+  
+  ```dart
+  import 'package:mockito/mockito.dart';
+
+class MockCalculator extends Mock implements Calculator {}
+
+// Create the mock object in your test case
+final calculator = MockCalculator();
+
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Fix this:

image

File: docs/developers/talawa/flutter_testing_guide.md [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+      await service.sendMessageToDirectChat(
+        id,
+        messageContent,
+      );
+    })
+```
+
+Here is a breakdown of what this test does
+
+1- The test starts by defining a mock object for the `_dbFunctions` class using the when function from the `Mockito` package. The mock object is set up to return a `QueryResult` object that simulates the result of a GraphQL mutation when the `gqlAuthMutation` method is called with the correct query and variables.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: use:

1.
2.
3.
4.

The CSS will handle the rest

File: docs/developers/talawa/flutter_testing_guide.md [hunk 8]
```diff
@@ -0,0 +1,271 @@
+---
+id: flutter-testing
+title: Flutter Testing Guide
+---
+
+# Flutter Testing Guide
+
+## Introduction
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:7]: Remove heading

File: docs/developers/talawa/flutter_testing_guide.md [hunk 9]
```diff
@@ -0,0 +1,271 @@
...
+
+Tests are an essential part of software development. They help developers to verify the functionality of the code they write and ensure that it behaves as expected. Testing is a process of executing a program with the intent of finding errors. In Flutter, testing is an essential part of the development process, and it's critical to write high-quality tests that cover all corner cases.
+
+
+## Why do we need tests?
+
+Tests ensure that the code you write works as intended. They help to catch bugs early in the development process, which saves time and effort in the long run. Writing tests also makes the code more reliable and maintainable.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Merge this section into that above. It is repetitive

**Comment by ALIRAZA17** [comment]
> Fix this:
> 
> image

@palisadoes Done!!

File: docs/developers/talawa/flutter_testing_guide.md [hunk 10]
```diff
@@ -0,0 +1,271 @@
...
+
+
+## How to cover all the corner cases?
+
+To ensure that your tests cover all corner cases, you should consider the different input values and edge cases that your code might encounter. For example, if you are testing a function that performs a calculation, you should test it with different input values, including negative numbers, zero, and large numbers.
+
+It's also a good idea to use boundary testing, where you test the boundaries between different input values. For example, if your function takes an input between 0 and 100, you should test it with values of 0, 1, 99, 100, and values just above and below these boundaries.
+
+
+## How to start with Flutter Testing?
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to "Getting Started"

File: docs/developers/talawa/flutter_testing_guide.md [hunk 11]
```diff
@@ -0,0 +1,271 @@
...
+
+1. Add the `flutter_test` package to your `pubspec.yaml` file.
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:27]: Add an introductory sentence before bullets. It makes the flow better

File: docs/developers/talawa/flutter_testing_guide.md [hunk 12]
```diff
@@ -0,0 +1,271 @@
...
+ 
+2. Create a new test file in your project's `test` directory. The file should have the same name as the file you want to test, with `_test` appended to the end. For example, if you want to test a file called `my_widget.dart`, the test file should be called `my_widget_test.dart`.
+ 
+3. Write test cases for the functions, widgets, or other parts of your application that you want to test. Use the tools provided by the flutter_test package, such as the test() and expect() functions, to define your test cases.
+ 
+4. Run your tests using the flutter test command. This will run all the tests in your project's test directory.
+
+### Basic Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to "Basic Test Example", otherwise the question is Example of what? The titles need to be self explanatory.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 13]
```diff
@@ -0,0 +1,271 @@
...
+
+
+## Why do we need tests?
+
+Tests ensure that the code you write works as intended. They help to catch bugs early in the development process, which saves time and effort in the long run. Writing tests also makes the code more reliable and maintainable. 
+
+The cost of removing defects increases exponentially. A defect caught in requirement and design phase costs less to fix than an error caught in the software maintenance cycle.
+
+
+## How to cover all the corner cases?
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to "Understanding Corner Cases"

File: docs/developers/talawa/flutter_testing_guide.md [hunk 14]
```diff
@@ -0,0 +1,271 @@
...
+
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:17]: Add a brief sentence about what a corner case is. Don't expect everyone to know.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 15]
```diff
@@ -0,0 +1,271 @@
...
+This will run all the tests in your project, including the `Calculator add` method test we just wrote. If the test passes, you should see the following output in your terminal:
+
+```bash
+00:00 +1: All tests passed!
+```
+If the test fails (for example, if the add method in calculator.dart was implemented incorrectly), you will see an error message in your terminal indicating what went wrong.
+
+
+## What are mocks and how to generate them?
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: 1. Create  a title "In Depth Example"
2. Add an introductory sentence
3. Make the mocks, stubs and test sections fall under it

Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to just "Mocks". Some of the titles are so long that they wrap twice in the right hand side bar.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 16]
```diff
@@ -0,0 +1,271 @@
...
+});
+  ```
+  
+  In this test, the `MockCalculator` is created and passed to the `CalculatorController`. The `when` method is used to stub the `add` method on the mock calculator, so
+  that when the `add` method is called with arguments 2 and 3, it returns the value 5. Then, the `addNumbers` method is called on the `CalculatorController`, and the
+  result is verified using the `expect` method. Finally, the `verify` method is used to ensure that the `add` method was called on the mock calculator with the correct
+  arguments.
+
+
+## How to Stub Methods and Classes?
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to "Stubs" Add another #

File: docs/developers/talawa/flutter_testing_guide.md [hunk 17]
```diff
@@ -0,0 +1,271 @@
...
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return 'my result', regardless of the input value.
+
+## Talawa App Tests Example on How to use Mocks and Stubs Correctly
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to "Example Test  of Mocks and Stubs" add #

File: docs/developers/talawa/flutter_testing_guide.md [hunk 18]
```diff
@@ -0,0 +1,271 @@
...
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests have all the necessary mocks and stubs to understand the concept of testing in talawa.
+
+### sendMessageToDirectChat Method
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: 1. Add # 
2. The title doesn't explain what you are trying to do.
3. Add some introductory text below the title. It makes the flow better. You have a title and then code.
4. If this is the code to be tested then say so. It's not clear.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 19]
```diff
@@ -0,0 +1,271 @@
...
+      result.data['sendMessageToDirectChat'] as Map<String, dynamic>,
+    );
+
+    _chatMessageController.add(message);
+
+    debugPrint(result.data.toString());
+  }
+```
+
+Test written for this method looks like this
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:203]: There should be a title here like "Sample Mock and Test Code" with the text below.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 20]
```diff
@@ -0,0 +1,271 @@
...
+      );
+      final service = ChatService();
+      await service.sendMessageToDirectChat(
+        id,
+        messageContent,
+      );
+    })
+```
+
+Here is a breakdown of what this test does
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:248]: There should be a title here too. The segmentation makes the readability better.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 21]
```diff
@@ -0,0 +1,271 @@
...
+2- The `ChatService` class is instantiated, and the `sendMessageToDirectChat` method is called with the correct `chatId` and `messageContent` parameters.
+
+3- Finally, the test verifies that the `_chatMessageController` object has been updated with the correct `ChatMessage` object that was received from the mocked `GraphQL mutation` result.
+
+4- The `when` function is used to set up a mock behavior for the `gqlAuthMutation` method of the `_dbFunctions` object. The mocked behavior returns a `QueryResult` object that simulates the result of a GraphQL mutation. The `QueryResult` object contains a map with a key of 'sendMessageToDirectChat', which contains a value that represents the returned ChatMessage object from the mutation.
+
+Overall, this test verifies that the sendMessageToDirectChat method correctly triggers a GraphQL mutation and correctly handles the returned data by updating the _chatMessageController object with the expected ChatMessage object.
+
+
+## What to do if you find a bug while writing a test for that file?
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change this to "Toubleshooting"

**Review by palisadoes — CHANGES_REQUESTED** [review]
1. Please make the suggested changes. 
2. The flow will be much better with them.
3. Please make sure the menuing on the right hand side makes sense when you are done.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 22]
```diff
@@ -0,0 +1,268 @@
...
+    ```
+ 3. Define a mock object by extending the Mock class:
+ 
+    ```dart
+    class MockMyObject extends Mock implements MyObject {}
+    ```
+ 4. Use the mock object in your test cases:
+
+    ```dart
+    test('my test case', () {
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Fix

image

**Review by palisadoes — CHANGES_REQUESTED** [review]

File: docs/developers/talawa/flutter_testing_guide.md (hunk 20 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md:248]: What should be the title here?

Comment (palisadoes) [reply docs/developers/talawa/flutter_testing_guide.md:248]: Think of something appropriate given my prior advice

File: docs/developers/talawa/flutter_testing_guide.md (hunk 5 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: I couldn't get it?

File: docs/developers/talawa/flutter_testing_guide.md (hunk 8 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md:7]: which one

Comment (palisadoes) [reply docs/developers/talawa/flutter_testing_guide.md:7]: Flutter Testing Guide

File: docs/developers/talawa/flutter_testing_guide.md (hunk 15 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: I should create this heading under the heading of what are mocks and how to generate them?

File: docs/developers/talawa/flutter_testing_guide.md (hunk 17 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: '#' should be single or added into the previous '##'?

File: docs/developers/talawa/flutter_testing_guide.md (hunk 18 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: @palisadoes Im a bit confused here. Do I need to add a single '#' here?

**Comment by ALIRAZA17** [comment]
@palisadoes I have commited again with all the changes you requested. I was confused about the used of '#' on certain heading so if you review them again to see if there are any mistakes left?

File: docs/developers/talawa/flutter_testing_guide.md (hunk 1 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: image

File: docs/developers/talawa/flutter_testing_guide.md (hunk 3 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: image

File: docs/developers/talawa/flutter_testing_guide.md (hunk 4 above)
Comment (ALIRAZA17) [reply docs/developers/talawa/flutter_testing_guide.md]: image

File: docs/developers/talawa/flutter_testing_guide.md [hunk 23]
```diff
@@ -0,0 +1,274 @@
...
+
+1. Add the `flutter_test` package to your `pubspec.yaml` file.
+ 
+2. Create a new test file in your project's `test` directory. The file should have the same name as the file you want to test, with `_test` appended to the end. For example, if you want to test a file called `my_widget.dart`, the test file should be called `my_widget_test.dart`.
+ 
+3. Write test cases for the functions, widgets, or other parts of your application that you want to test. Use the tools provided by the flutter_test package, such as the test() and expect() functions, to define your test cases.
+ 
+4. Run your tests using the flutter test command. This will run all the tests in your project's test directory.
+
+### Basic Test Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Remove a #

File: docs/developers/talawa/flutter_testing_guide.md [hunk 24]
```diff
@@ -0,0 +1,274 @@
...
+```
+
+This will run all the tests in your project, including the `Calculator add` method test we just wrote. If the test passes, you should see the following output in your terminal:
+
+```bash
+00:00 +1: All tests passed!
+```
+If the test fails (for example, if the add method in calculator.dart was implemented incorrectly), you will see an error message in your terminal indicating what went wrong.
+
+### In Depth Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: 1. Remove a #
2. Make it Examples

File: docs/developers/talawa/flutter_testing_guide.md [hunk 25]
```diff
@@ -0,0 +1,274 @@
...
+    class MockMyObject extends Mock implements MyObject {}
+    ```
+ 4. Use the mock object in your test cases:
+    
+  ```dart
+  class MockMyObject extends Mock implements MyObject {}  
+  ```
+
+
+### Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: #### Mocks Test Example

File: docs/developers/talawa/flutter_testing_guide.md [hunk 26]
```diff
@@ -0,0 +1,274 @@
...
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return `my result`, regardless of the input value.
+
+# Example Test of Mocks and Stubs
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: #### Mocks and Stubs Test Example

File: docs/developers/talawa/flutter_testing_guide.md [hunk 27]
```diff
@@ -0,0 +1,274 @@
...
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests have all the necessary mocks and stubs to understand the concept of testing in talawa.
+
+# sendMessageToDirectChat Method
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: ##### "Method Under Test"

File: docs/developers/talawa/flutter_testing_guide.md [hunk 28]
```diff
@@ -0,0 +1,274 @@
...
+      result.data['sendMessageToDirectChat'] as Map<String, dynamic>,
+    );
+
+    _chatMessageController.add(message);
+
+    debugPrint(result.data.toString());
+  }
+```
+
+# Sample Mock and Test Code
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to ##### Sample Mock and Test Code

File: docs/developers/talawa/flutter_testing_guide.md [hunk 29]
```diff
@@ -0,0 +1,274 @@
...
+      );
+      final service = ChatService();
+      await service.sendMessageToDirectChat(
+        id,
+        messageContent,
+      );
+    })
+```
+
+# Test Explanation
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: Change to ##### Test Explanation

File: docs/developers/talawa/flutter_testing_guide.md [hunk 30]
```diff
@@ -0,0 +1,274 @@
...
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return `my result`, regardless of the input value.
+
+# Example Test of Mocks and Stubs
+
+There is a function in talawa which provides chat service in the app. The file is located in `talawa/lib/services/chat_service.dart`. Unit Tests for this file have been written to test all the methods written in the file for instance `sendMessageToDirectChat` and `getDirectChatsByUserId` methods. These tests are written in the file `talawa\test\service_tests\chat_service_test.dart`. These tests have all the necessary mocks and stubs to understand the concept of testing in talawa.
+
+# sendMessageToDirectChat Method
+
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md:177]: Add introductory text that states that in this example we'll be testing the `sendMessageToDirectChat` method. It makes the flow better.

**Review by palisadoes — CHANGES_REQUESTED** [review]
Please make the requested changes.

**Comment by ALIRAZA17** [comment]
@palisadoes Done!!

File: docs/developers/talawa/flutter_testing_guide.md [hunk 31]
```diff
@@ -0,0 +1,274 @@
...
+`when()` method is used in the previous Calculator example as well where it is used to stub the add method to mock the Calculator.
+
+You can also use the any matcher to match any input value. For example:
+
+```dart
+when(myMock.myMethod(any)).thenReturn('my result');
+```
+This will stub the `myMethod()` method to always return `my result`, regardless of the input value.
+
+# Mocks and Stubs Test Example
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: This should be:

`### Mocks and Stubs Test Example`

The title doesn't make sense as a main heading if it's under Stubs.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 32]
```diff
@@ -0,0 +1,274 @@
...
+
+In this example, we'll be testing the `sendMessageToDirectChat` method from our application. This method is responsible for sending direct messages between two users in a private chat. The `sendMessageToDirectChat` method is critical to the functionality of our application, and we need to ensure that it works correctly under a variety of conditions. To do so, we'll be using a combination of manual and automated testing techniques to thoroughly test this method and uncover any potential bugs or issues. By the end of this example, you'll have a better understanding of how to approach testing for critical methods in this application. The file is located in `talawa/lib/services/chat_service.dart` and its tests are written in the file `talawa\test\service_tests\chat_service_test.dart`
+
+# Method Under Test
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: This should be:

`#### Method Under Test`

The title doesn't make sense as a main heading if it's under Stubs.

File: docs/developers/talawa/flutter_testing_guide.md [hunk 33]
```diff
@@ -0,0 +1,274 @@
...
+      result.data['sendMessageToDirectChat'] as Map<String, dynamic>,
+    );
+
+    _chatMessageController.add(message);
+
+    debugPrint(result.data.toString());
+  }
+```
+
+##### Sample Mock and Test Code
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: This should be:

`#### Sample Mock and Test Code`

File: docs/developers/talawa/flutter_testing_guide.md [hunk 34]
```diff
@@ -0,0 +1,274 @@
...
+      );
+      final service = ChatService();
+      await service.sendMessageToDirectChat(
+        id,
+        messageContent,
+      );
+    })
+```
+
+##### Test Explanation
```
Comment (palisadoes) [inline docs/developers/talawa/flutter_testing_guide.md]: `#### Test Explanation`

**Review by palisadoes — CHANGES_REQUESTED** [review]
That should be it.

**Comment by ALIRAZA17** [comment]
@palisadoes Done!!

**Review by palisadoes — APPROVED** [review]

**Comment by literalEval** [comment]
@ALIRAZA17 please have a closer look at the guide in the coming time and any feedback it gets (especially from newcomers). If anyone suggests any additions, please think about them carefully. If a beginner learns something from your guide, then only your work for it will be successful :)

Suggesting this as a fellow contributor :)

**Comment by ALIRAZA17** [comment]
> @ALIRAZA17 please have a closer look at the guide in the coming time and any feedback it gets (especially from newcomers). If anyone suggests any additions, please think about them carefully. If a beginner learns something from your guide, then only your work for it will be successful :)
> 
> Suggesting this as a fellow contributor :)

Indeed I will keep an eye on this. Thanks for the input!!
